mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp

```diff
@@ -115,17 +116,88 @@ void AffixTuningParameters::affixTuningParametersImpl(T &op) {
     }
   }
 
+  return std::make_tuple(y, x, ho, wo, hi, wi, k, c, n);
+}
+
+void AffixTuningParameters::affixBackwardWeightUtilityKernels(
+    miopen::Conv2DBwdWeightOp &op) {
+  auto gemmIdAttr = op->template getAttrOfType<IntegerAttr>("gemm_id");
+  assert(gemmIdAttr);
+  int64_t gemmId = gemmIdAttr.getInt();
+
+  auto xdlopsV2Attr = op->template getAttrOfType<BoolAttr>("xdlopsV2");
+  if (xdlopsV2Attr && xdlopsV2Attr.getValue() == true) {
+    OpBuilder b(op.getContext());
+
+    ConvolutionContext convContext = populateConvContext(op);
+
+    // get y, x, ho, wo, hi, wi, k, c, n
+    int64_t y, x, ho, wo, hi, wi, k, c, n;
+    std::tie(y, x, ho, wo, hi, wi, k, c, n) = fetchDimensions(op);
+
+    int64_t gemmMSize, gemmNSize, gemmKSize;
+    gemmMSize = k;
+    gemmKSize = n * ho * wo;
+    gemmNSize = c * y * x;
+
+    int64_t gemmMExtra, gemmNExtra, gemmKExtra;
+    gemmMExtra = gemmNExtra = gemmKExtra = 0;
+
+    // isOriginalKernelSupport is not used.
+    // Only needExtraPad is used.
+    bool isOriginalKernelSupport = true;
+    bool needExtraPad = false;
+    PopulateParamsXDL populateParamsXDL;
+    std::tie(isOriginalKernelSupport, needExtraPad, gemmMExtra, gemmNExtra,
+             gemmKExtra) =
+        calculatePaddingKernelSize(
+            gemmMSize, gemmNSize, gemmKSize, convContext.getOpType(),
+            convContext.getDataType(), populateParamsXDL);
+
+    // For padding cases, gemmId must be 0.
+    if (needExtraPad == true) {
+      assert(gemmId == 0);
+    } else {
+      assert((gemmId >= 0) && (gemmId < 3));
+      switch (gemmId) {
+      case 0:
+      case 2:
+        // Override grid_size and block_size be 1 for utility kernels.
+        // FIXME. Use better sizes for speedups.
+        op->setAttr("grid_size", b.getI32IntegerAttr(1));
```

Comment:
What do grid_size and block_size do?

Reply:
These are the parameters sent to the runtime when we launch a kernel. It specifies the size of a workgroup (block_size), and how many workgroups (grid_size) we need. Here you can see we use an extremely conservative way to ease the programming of the code emission logic for those utility kernels.

Ideally, we want something like this:
```
    size_t offset = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x);
    size_t stride = hipBlockDim_x * hipGridDim_x;

    for (size_t i = offset; i < N; i += stride) {
       // do the elementwise operations we want
    }
```

I was hoping upstream transforms for linalg could help me achieve that but I couldn't find an end-to-end pipeline yet.

Reply:
I'd've gone for a different breakup.

```
per_thread_work = N / (n_total_threads)
offset = global_thread_id * per_thread_work
for (int i = 0; i < per_thread_work; ++i)
  buffer[offset + i] = 0;
```

Reply:
And our coordinate transform scheme should be able to support this?

Reply:
on another round of examining samples in upstream, this pipeline may offer some help:
```
-convert-linalg-to-parallel-loops -convert-parallel-loops-to-gpu
```
I'll do some local experiments. Should we get something working end-to-end I'll submit a new PR.

Reply:
tracking it at https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/466